Enhancement - add trace context fields, as custom fields

sgryphon opened the pull request:
This adds the span-id and parent-id distributed trace correlation fields as custom fields.

One of the fields, span.id, is included in ECS 1.6, and can simply be replaced when upgrading.

This implements it as a custom field, Span.id, using uppercase conventions as specified in the ECS documentation.

The Parent.id field is also included (this has been mentioned for ECS, but not yet included).

The values (TraceId, SpanId, and ParentId), based on W3C Trace Context, are support in .NET (via the Activity class), and automatically populated in modern ASP.NET applications (e.g. via the traceparent header). ASP.NET also sets a logging Scope with the three values.

--

This Pull Request is on top of the bug fix to get custom fields supported again, as while implementing I noticed that had been broken.  

It would be nice, but not essential, to merge this before Extensions GA #95.

apmmachine commented:
## :grey_exclamation: Build Aborted
  
  > The PR is not allowed to run in the CI yet
  


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fecs-dotnet-mbp%2FPR-115/detail/PR-115/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fecs-dotnet-mbp%2FPR-115/detail/PR-115/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fecs-dotnet-mbp%2FPR-115/detail/PR-115/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fecs-dotnet-mbp%2FPR-115/detail/PR-115/1//artifacts) 

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #115 opened]`

* Reason: `The PR is not allowed to run in the CI yet`

* Start Time: 2020-11-16T03:16:02.674+0000

* Duration: 2 min 59 sec


* Commit: 32fa6922fb4019c0249249f39fb14e3bd2579da5




</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-dotnet%2Fecs-dotnet-mbp%2FPR-115/detail/PR-115/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  
  
  
  
  
#### `Load a resource file from a shared library`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/ecs-dotnet-mbp/pipelines/PR-115/runs/1/steps/75/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/pipelines/ecs-dotnet-mbp/pipelines/PR-115/runs/1/steps/78/log/?start=0">here</a></li>
<li>Description: <code>githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permission</code></l1>
</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-11-16T03:17:32.187Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-11-16T03:17:32.215Z]  > git fetch --no-tags --progress -- git@github.com:elastic/ecs-dotnet.git +refs/heads/*:refs/remotes/origin/* # timeout=15
[2020-11-16T03:17:33.027Z]  > git config remote.origin.url git@github.com:elastic/ecs-dotnet.git # timeout=10
[2020-11-16T03:17:33.035Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-11-16T03:17:33.049Z]  > git config remote.origin.url git@github.com:elastic/ecs-dotnet.git # timeout=10
[2020-11-16T03:17:33.057Z]  > git rev-parse --verify HEAD # timeout=10
[2020-11-16T03:17:33.063Z] No valid HEAD. Skipping the resetting
[2020-11-16T03:17:33.063Z]  > git clean -fdx # timeout=10
[2020-11-16T03:17:33.079Z] Fetching upstream changes from git@github.com:elastic/ecs-dotnet.git
[2020-11-16T03:17:33.079Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-11-16T03:17:33.083Z]  > git fetch --no-tags --progress --prune -- git@github.com:elastic/ecs-dotnet.git +refs/pull/115/head:refs/remotes/origin/PR-115 +refs/heads/master:refs/remotes/origin/master # timeout=15
[2020-11-16T03:17:33.747Z] Merging remotes/origin/master commit 718a86a3d947dffa3058a35743afe87a4c90c641 into PR head commit 32fa6922fb4019c0249249f39fb14e3bd2579da5
[2020-11-16T03:17:33.890Z] Merge succeeded, producing 32fa6922fb4019c0249249f39fb14e3bd2579da5
[2020-11-16T03:17:33.891Z] Checking out Revision 32fa6922fb4019c0249249f39fb14e3bd2579da5 (PR-115)
[2020-11-16T03:17:33.763Z]  > git config core.sparsecheckout # timeout=10
[2020-11-16T03:17:33.768Z]  > git checkout -f 32fa6922fb4019c0249249f39fb14e3bd2579da5 # timeout=15
[2020-11-16T03:17:33.843Z]  > git remote # timeout=10
[2020-11-16T03:17:33.852Z]  > git config --get remote.origin.url # timeout=10
[2020-11-16T03:17:33.857Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-11-16T03:17:33.861Z]  > git merge 718a86a3d947dffa3058a35743afe87a4c90c641 # timeout=10
[2020-11-16T03:17:33.885Z]  > git rev-parse HEAD^{commit} # timeout=10
[2020-11-16T03:17:33.896Z]  > git config core.sparsecheckout # timeout=10
[2020-11-16T03:17:33.917Z]  > git checkout -f 32fa6922fb4019c0249249f39fb14e3bd2579da5 # timeout=15
[2020-11-16T03:17:37.995Z] Commit message: "Use custom fields for distributed trace context"
[2020-11-16T03:17:37.995Z] First time build. Skipping changelog.
[2020-11-16T03:17:37.995Z] Cleaning workspace
[2020-11-16T03:17:38.053Z]  > git --version # timeout=10
[2020-11-16T03:17:38.057Z]  > git --version # 'git version 2.17.1'
[2020-11-16T03:17:38.060Z] fatal: bad object 5ad02210061fc489e0a94d419f5ccbb76fc5b28b
[2020-11-16T03:17:37.998Z]  > git rev-parse --verify HEAD # timeout=10
[2020-11-16T03:17:38.003Z] Resetting working tree
[2020-11-16T03:17:38.003Z]  > git reset --hard # timeout=10
[2020-11-16T03:17:38.009Z]  > git clean -fdx # timeout=10
[2020-11-16T03:17:38.861Z] Masking supported pattern matches of $JOB_GCS_BUCKET or $NOTIFY_TO
[2020-11-16T03:17:38.937Z] Timeout set to expire in 1 hr 0 min
[2020-11-16T03:17:38.982Z] The timestamps step is unnecessary when timestamps are enabled for all Pipeline builds.
[2020-11-16T03:17:39.399Z] [INFO] Number of builds to be searched 10
[2020-11-16T03:17:39.618Z] [INFO] 'shallow' is forced to be disabled when running on PullRequests
[2020-11-16T03:17:39.653Z] Running in /var/lib/jenkins/workspace/ent-dotnet_ecs-dotnet-mbp_PR-115/src/go.elastic.co/apm/ecs-dotnet
[2020-11-16T03:17:39.698Z] [INFO] gitCheckout: Checkout SCM PR-115 with some customisation.
[2020-11-16T03:17:39.747Z] [INFO] Override default checkout
[2020-11-16T03:17:39.814Z] Sleeping for 10 sec
[2020-11-16T03:17:49.930Z] using credential f6c7695a-671e-4f4f-a331-acdce44ff9ba
[2020-11-16T03:17:49.991Z] Cloning the remote Git repository
[2020-11-16T03:17:50.005Z] Cloning repository git@github.com:elastic/ecs-dotnet.git
[2020-11-16T03:17:50.042Z]  > git init /var/lib/jenkins/workspace/ent-dotnet_ecs-dotnet-mbp_PR-115/src/go.elastic.co/apm/ecs-dotnet # timeout=10
[2020-11-16T03:17:50.062Z] Fetching upstream changes from git@github.com:elastic/ecs-dotnet.git
[2020-11-16T03:17:50.062Z]  > git --version # timeout=10
[2020-11-16T03:17:50.071Z]  > git --version # 'git version 2.17.1'
[2020-11-16T03:17:50.072Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-11-16T03:17:50.077Z]  > git fetch --tags --progress -- git@github.com:elastic/ecs-dotnet.git +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-11-16T03:17:51.587Z] Checking out Revision 32fa6922fb4019c0249249f39fb14e3bd2579da5 (origin/PR-115)
[2020-11-16T03:17:51.653Z] Commit message: "Use custom fields for distributed trace context"
[2020-11-16T03:17:51.665Z]  > git --version # timeout=10
[2020-11-16T03:17:51.677Z]  > git --version # 'git version 2.17.1'
[2020-11-16T03:17:51.683Z] fatal: bad object 5ad02210061fc489e0a94d419f5ccbb76fc5b28b
[2020-11-16T03:17:50.862Z]  > git config remote.origin.url git@github.com:elastic/ecs-dotnet.git # timeout=10
[2020-11-16T03:17:50.865Z]  > git config --add remote.origin.fetch +refs/heads/*:refs/remotes/origin/* # timeout=10
[2020-11-16T03:17:50.871Z]  > git config remote.origin.url git@github.com:elastic/ecs-dotnet.git # timeout=10
[2020-11-16T03:17:50.877Z] Fetching upstream changes from git@github.com:elastic/ecs-dotnet.git
[2020-11-16T03:17:50.878Z] using GIT_SSH to set credentials GitHub user @elasticmachine SSH key
[2020-11-16T03:17:50.885Z]  > git fetch --tags --progress -- git@github.com:elastic/ecs-dotnet.git +refs/pull/115/head:refs/remotes/origin/PR-115 +refs/heads/master:refs/remotes/origin/master # timeout=10
[2020-11-16T03:17:51.578Z]  > git rev-parse origin/PR-115^{commit} # timeout=10
[2020-11-16T03:17:51.590Z]  > git config core.sparsecheckout # timeout=10
[2020-11-16T03:17:51.606Z]  > git checkout -f 32fa6922fb4019c0249249f39fb14e3bd2579da5 # timeout=10
[2020-11-16T03:17:52.854Z] Masking supported pattern matches of $GIT_USERNAME or $GIT_PASSWORD
[2020-11-16T03:17:53.451Z] + git fetch https://****:****@github.com/elastic/ecs-dotnet.git +refs/pull/*/head:refs/remotes/origin/pr/*
[2020-11-16T03:17:54.060Z] Archiving artifacts
[2020-11-16T03:17:54.655Z] + git rev-parse HEAD
[2020-11-16T03:17:55.011Z] + git rev-parse HEAD
[2020-11-16T03:17:55.314Z] + git rev-parse origin/pr/115
[2020-11-16T03:17:55.364Z] [INFO] githubEnv: Found Git Build Cause: pr
[2020-11-16T03:17:55.615Z] Masking supported pattern matches of $GITHUB_TOKEN
[2020-11-16T03:17:56.734Z] [WARN] githubApiCall: The REST API call https://api.github.com/repos/elastic/ecs-dotnet/pulls/115/reviews return 0 elements
[2020-11-16T03:17:56.788Z] [INFO] githubPrCheckApproved: Title: Enhancement - add trace context fields, as custom fields - User: sgryphon - Author Association: CONTRIBUTOR
[2020-11-16T03:17:57.095Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet
[2020-11-16T03:17:57.095Z] ERROR: githubPrCheckApproved: The PR is not allowed to run in the CI yet. (Only users with write permissions can do so.)
[2020-11-16T03:17:57.163Z] [INFO] Let's stop build #1. The PR is not allowed to run in the CI yet
[2020-11-16T03:17:57.183Z] Sleeping for 5 sec
[2020-11-16T03:17:58.257Z] Stage "Sanity checks" skipped due to earlier failure(s)
[2020-11-16T03:17:58.350Z] Stage "Parallel" skipped due to earlier failure(s)
[2020-11-16T03:17:58.448Z] Stage "Linux" skipped due to earlier failure(s)
[2020-11-16T03:17:58.460Z] Stage "Windows" skipped due to earlier failure(s)
[2020-11-16T03:17:58.493Z] Failed in branch Linux
[2020-11-16T03:17:58.533Z] Stage "Windows" skipped due to earlier failure(s)
[2020-11-16T03:17:58.630Z] Failed in branch Windows
[2020-11-16T03:17:58.747Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-16T03:17:58.778Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-16T03:17:58.847Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-16T03:17:58.925Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-16T03:17:59.003Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-16T03:18:00.239Z] Running on worker-1225339 in /var/lib/jenkins/workspace/ent-dotnet_ecs-dotnet-mbp_PR-115
[2020-11-16T03:18:00.351Z] [INFO] getVaultSecret: Getting secrets
[2020-11-16T03:18:00.537Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-11-16T03:18:02.763Z] + chmod 755 generate-build-data.sh
[2020-11-16T03:18:02.763Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/ecs-dotnet-mbp/PR-115/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/ecs-dotnet-mbp/PR-115/runs/1 ABORTED 118594
[2020-11-16T03:18:02.763Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/ecs-dotnet-mbp/PR-115/runs/1/steps/?limit=10000 -o steps-info.json
[2020-11-16T03:18:05.013Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-dotnet/ecs-dotnet-mbp/PR-115/runs/1/tests/?status=FAILED -o tests-errors.json
[2020-11-16T03:18:05.014Z] Retry 1/3 exited 22, retrying in 1 seconds...
[2020-11-16T03:18:07.258Z] Retry 2/3 exited 22, retrying in 2 seconds...

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

Mpdreamz commented:
@sgryphon mind rebasing this against master now that #107 is finally merged, thanks!

sgryphon commented:
TL; DR; this is wrong, I'm going to change to lower case.

Longer discussion:

I was thinking about compatibility, and this is actually the wrong thing to do. If we add custom fields with upper case, e.g. Span.id, then when the real span.id (ECS 1.6) is implemented we face a breaking change by removing Span.id (or, probably worse, supporting both).

More broader, for compatibility: if we are running logger 1.5, and pushing to a system that supports ECS 1.4, then backwards compatibility should mean we are okay (as it is only a minor version difference). We are pushing extra fields, but they would be ignored (like custom fields). The other side is only going to report / use the 1.4 fields.

In the reverse direction, if we are running logger 1.5, and push to a system that wants ECS 1.7, then it also should not be a problem. We can't trigger anything based on a 1.7 field (e.g. email alert), because we don't send them. But it should be able to at least handle the missing field (as fields can be blank).

One of the reasons I was thinking about this was whether we needed to have one release of the logger support potentially configurable versions of ECS, e.g. select which version of ECS to use. Seeing as versions are backwards compatible this doesn't make sense. It is the first scenario above. There is not point having the ability to downgrade what we sent to 1.4; we just sent 1.5 and the target system will ignore the extra fields.

So, for span.id, we also have the situation that implementing new fields that match those in future versions is also not a problem.

If we release version 1.5.1, that adds span.id (lowercase), then 1.4 destinations will continue to ignore it, and 1.7 destinations will start using it, even though we aren't filling in other 1.7 fields (they just assume the fields are blank).

In fact, the whole notion of supporting 1.5 is a bit weird for the logger; sure the ECS library has placeholders for all the 1.5 fields, but none of the loggers fill in every single field, or even have the potential to fill in all fields (there are many that are simply ignored).

From that point of view the logger could simply claim it is 1.7, just not sending any of the new fields. Although I agree the actual library itself should have every 1.7 field before it claims it supports version 1.7.

But we could still add partial support, e.g. 1.5.1 with the span.id field, and then bump to claim 1.6 once the rest are added.

Note that for fields that really are custom, like Scope, and won't be part of ECS, using title case is correct.

So, for span.id, this is fine, as it will be in a future version.

parent.id is a bit more tricky.   I am going to put up a proposal to add it to ECS, which probably wouldn't be until at least 1.9, and I'm not even sure it would be approved (there was a discussion on an issue).

Maybe the best is to leave parent.id until the change to ECS is either approved or rejected. If approved add it as parent.id, for future compatibility.

If not, then consider adding Parent.id (like Scope, etc). If these ever got added, e.g. scope got added, then we would have to consider duplicating vs breaking change.

sgryphon commented:
Updated as per my longer comments to be forward compatible with ECS 1.6, as we know the field span.id is in the newer versions. Removed parent.id (until it appears in the standard). 

Also updated for the changes in the custom field fix branch, to support deserialisation as well. i.e. this PR includes everything the bug fix PR does, plus the span.id extra bit.

sgryphon commented:
@russcam I have fixed the merge conflicts on this one as well, if you can review please.

russcam commented:
@sgryphon I'm in the process of updating to ECS 1.6.0 version (and to follow with subsequent versions). I hope to have this ready to do a release very soon, including GA'ing the great work you've put into Elasticsearch.Extensions.Logging.

[Since 1.6 onwards has the `span.id`](https://www.elastic.co/guide/en/ecs/1.6/ecs-tracing.html), and with the plan to release 1.6.0 as soon as possible, I'm thinking this PR might be moot. What are your thoughts?

sgryphon commented:
We still need the code to pull out span.id and insert it into the log event, i.e. the code in ExtractW3cSpanIdFromActivityId and AddTracing. Even the line logEvent.Span.Id would be the same.

What would change would be Span property on LogEvent, and the custom serialisation, which would be replaced by the same property on Base. It may actually still work, as LogEvent.Span will simply hide the inherited member Base.Span (compiler warnings, etc).

So, you still want to merge to get all the rest of the code, then when combining with the 1.6 branch simply remove LogEvent.Span and recompile.

sgryphon commented:
@russcam I'm happy to do the patch to version 1.6, but half of this PR is still required (to get the value), just the setting half changes. So the PR is still required (even for 1.6)

russcam commented:
@sgryphon I'm about to open a PR that updates to ECS 1.6.0 👍 

Pulling the trace data out of `Activity` I think is a reasonable default. I think it should be possible to override the default to pull trace data out of Elastic APM's transaction, if present, similar to [Elastic.Apm.SerilogEnricher](https://github.com/elastic/ecs-dotnet/tree/master/src/Elastic.Apm.SerilogEnricher) and [Elastic.Apm.NLog](https://github.com/elastic/ecs-dotnet/tree/master/src/Elastic.Apm.NLog)

sgryphon commented:
Sounds like a good idea; I'll have a look at the other code. Are you okay if I do that as a separate pull request?

P.S. Pulling it out of Activity.Id is for backwards compatibility so it can be compiled with Standard 2.0, and work correctly with later versions. If we want to do multiple targeting for the library we can add conditional compilation for later version to just grab details directly. (Again, as a separate pull request though).

russcam commented:
Separate PR is good; this one already contains a monster amount of your effort, which is hugely appreciated 😃

russcam commented on src/Elasticsearch.Extensions.Logging/ElasticsearchLogger.cs:
```diff
@@ -129,9 +129,18 @@ private void AddTracing(LogEvent logEvent)
 				logEvent.Trace = new Trace() { Id = activity.RootId };
+
+				// Use field span.id forward compatible with ECS 1.6
+				var spanId = ExtractW3cSpanIdFromActivityId(activity.Id);
```
I think we could check if the `Activity` is not null and `activity.IdFormat == ActivityIdFormat.W3C`, and pull the trace id and span id from `activity.TraceId.ToHexString()` and `activity.SpanId.ToHexString()`, respectively. This would remove the need to do the W3C Trace Context parsing. If there is a desire to also handle the older hierarchical format, that could be handled in a separate if/else branch

sgryphon replied:
Activity.IdFormat, Activity.Span, etc are all netcore 3.0 features, and not available in netstandard2.0; hence the clunky pulling Activity.Id and seeing if the string matches the expected format (because the library is being used in a 3.0+ application).

We could do multi-targeting for the library packages, and then conditional compilation #if version >= 3.0 then grab SpanId, but I suggest maybe as a separate pull request.

russcam replied:
We're [using them in `Elastic.Apm`](https://github.com/elastic/apm-agent-dotnet/blob/c061236367df7b12a16e8e75d9d0abe12db34dfa/src/Elastic.Apm/Model/Transaction.cs#L183), which [targets netstandard2.0](https://github.com/elastic/apm-agent-dotnet/blob/c061236367df7b12a16e8e75d9d0abe12db34dfa/src/Elastic.Apm/Elastic.Apm.csproj#L3). Am I misunderstanding?

sgryphon replied:
Okay, so Elasticsearch.Net/7.6.1 has a dependency on System.Diagnostics.DiagnosticSource/5.0.0; which from looking is not actually part of netstandard proper, just a library built on top of it.

The documentation says you need Core 3.0 or above, https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.idformat

But that's not actually correct, all you need is a nuget reference to System.Diagnostics.DiagnosticSource, which runs on top of netstandard 1.1 or above, https://www.nuget.org/packages/System.Diagnostics.DiagnosticSource/5.0.0

It seems to all compile.

sgryphon replied:
Updated to use Activity.Trace/.Span, if W3C format, otherwise fallback (similar to older fallback to TraceContext.Activity)

russcam commented:
I'm about to push a commit to this branch that updates in line with ECS 1.6 changes

sgryphon commented:
> I'm about to push a commit to this branch that updates in line with ECS 1.6 changes

Yes. Just removing the custom span and using base; the rest of the code (to set the value) is the same.
